src/Microsoft.DotNet.CodeFormatting/Rules/HasNewLineAtEndOfFileFormattingRule.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.DotNet.CodeFormatting.Rules
+{
+    [SyntaxRuleOrder(SyntaxRuleOrder.HasNewLineAtEndOfFile)]
+    internal sealed class HasNewLineAtEndOfFileFormattingRule : ISyntaxFormattingRule
+    {
+        public SyntaxNode Process(SyntaxNode syntaxRoot)
+        {
+            var node = syntaxRoot.DescendantNodes().LastOrDefault();
+
+            if (node != null)
+            {
+                syntaxRoot = syntaxRoot.ReplaceNode(node, RemoveNewLines(node));
+            }
+
+            var token = syntaxRoot.DescendantTokens().Single(x => x.IsKind(SyntaxKind.EndOfFileToken));
```

Comment:
CC @jasonmalinowski because he will have better knowledge of the parse tree here.  My concern is whether or not we are guaranteed the `EndOfFileToken` will be the only descendant token here in all cases.  If not `Single` will throw and break the formatter.